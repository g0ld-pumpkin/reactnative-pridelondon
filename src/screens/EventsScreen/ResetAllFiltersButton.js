// @flow
import React from "react";
import { Animated, StyleSheet, Easing, InteractionManager } from "react-native";
import type { ViewLayoutEvent } from "react-native/Libraries/Components/View/ViewPropTypes";
import FilterHeaderButton from "./FilterHeaderButton";
import text from "../../constants/text";

type Props = {
  visible: boolean,
  onPress: () => void,
  animationTime: number,
  animationDelay: number
};

type State = {
  isAnimating: boolean
};

const DEFAULT_HEIGHT = 120;
const DEFAULT_FADE_VALUE = 1;
const DEFAULT_TOP_OFFSET_VALUE = 0;

class ResetAllFiltersButton extends React.PureComponent<Props, State> {
  static defaultProps = {
    animationTime: 300,
    animationDelay: 0
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      isAnimating: false
    };
  }

  componentWillUnmount() {
    this.animationFinished();
  }

  setButtonHeight = (e: ViewLayoutEvent): void => {
    const { height } = e.nativeEvent.layout;
    this.height = height;
  };

  fadeOut = (): void => {
    const { animationTime, animationDelay } = this.props;

    Animated.parallel([
      Animated.timing(this.fadeValue, {
        toValue: 0,
        duration: animationTime / 2,
        delay: animationDelay
      }),
      Animated.timing(this.topOffset, {
        toValue: -this.height,
        duration: animationTime,
        easing: Easing.out(Easing.quad),
        delay: animationDelay
      })
    ]).start(this.animationFinished);
  };

  resetAllFilters = (): void => {
    this.props.onPress();
    InteractionManager.runAfterInteractions(() => {
      this.setState({ isAnimating: true });
      this.fadeOut();
    });
  };

  resetAnimation = (): void => {
    this.topOffset = new Animated.Value(DEFAULT_TOP_OFFSET_VALUE);
    this.fadeValue = new Animated.Value(DEFAULT_FADE_VALUE);
  };

  animationFinished = (): void => {
    if (this.state.isAnimating) {
      this.setState({ isAnimating: false });
      this.resetAnimation();
    }
  };

  topOffset: Animated.Value = new Animated.Value(DEFAULT_TOP_OFFSET_VALUE);
  fadeValue: Animated.Value = new Animated.Value(DEFAULT_FADE_VALUE);
  height: number = DEFAULT_HEIGHT;

  render() {
    const isVisible: boolean = this.state.isAnimating || this.props.visible;
    return (
      isVisible && (
        <Animated.View
          onLayout={this.setButtonHeight}
          style={[
            styles.clearAllWrapper,
            { opacity: this.fadeValue, marginTop: this.topOffset }
          ]}
        >
          <FilterHeaderButton
            active={false}
            text={text.resetAllFilters}
            label={text.resetAllFilters}
            style={styles.clearAll}
            onPress={this.resetAllFilters}
          />
        </Animated.View>
      )
    );
  }
}

const styles = StyleSheet.create({
  clearAll: {
    minHeight: 0,
    paddingTop: 16
  },
  clearAllWrapper: {
    flexDirection: "row",
    justifyContent: "flex-end"
  }
});

export default ResetAllFiltersButton;
