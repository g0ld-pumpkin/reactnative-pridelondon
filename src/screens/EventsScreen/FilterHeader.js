// @flow
import React from "react";
import { View, StyleSheet } from "react-native";
import type { DateRange } from "../../data/date-time";
import type { EventCategoryName } from "../../data/event";
import FilterHeaderButton from "./FilterHeaderButton";
import ContentPadding from "../../components/ContentPadding";
import FilterHeaderCategories from "./FilterHeaderCategories";
import {
  filterBgColor,
  filterButtonsBgColor,
  whiteColor
} from "../../constants/colors";
import text from "../../constants/text";
import { formatDateRange } from "../../data/formatters";
import ResetAllFiltersButton from "./ResetAllFiltersButton";

export type Props = {
  onFilterCategoriesPress: Function,
  dateFilter: ?DateRange,
  onFilterButtonPress: () => void,
  onDateFilterButtonPress: () => void,
  selectedCategories: Set<EventCategoryName>,
  numTagFiltersSelected: number,
  resetAllFiltersPress: () => void,
  scrollEventListToTop: () => void
};

class FilterHeader extends React.PureComponent<Props> {
  static defaultProps = {
    resetAllFiltersPress: () => {}
  };

  resetAllFilters = () => {
    this.props.resetAllFiltersPress();
    this.props.scrollEventListToTop();
  };

  render() {
    const {
      dateFilter,
      onFilterCategoriesPress,
      selectedCategories,
      onFilterButtonPress,
      onDateFilterButtonPress,
      numTagFiltersSelected
    } = this.props;
    const formattedDateFilter = dateFilter
      ? formatDateRange(dateFilter)
      : text.selectDates;

    const anyAppliedFilters: boolean =
      !!dateFilter || numTagFiltersSelected > 0 || selectedCategories.size > 0;

    return (
      <View accessibilityTraits={["header"]} style={styles.container}>
        <ContentPadding>
          <View>
            <ResetAllFiltersButton
              visible={anyAppliedFilters}
              onPress={this.resetAllFilters}
            />
            <View testID="event-filter-header" style={styles.content}>
              <FilterHeaderCategories
                onFilterPress={onFilterCategoriesPress}
                selectedCategories={selectedCategories}
              />
            </View>
          </View>
        </ContentPadding>
        <View style={styles.contentFilters}>
          <FilterHeaderButton
            active={!!dateFilter}
            text={formattedDateFilter}
            label={`filter by date: ${formattedDateFilter}`}
            onPress={onDateFilterButtonPress}
            style={styles.filterButton}
            testID="open-date-filters-button"
          />
          <View style={styles.dividerLine} />
          <FilterHeaderButton
            active={numTagFiltersSelected > 0}
            text={numTagFiltersSelected > 0 ? text.filters : text.addFilters}
            label={numTagFiltersSelected > 0 ? text.filters : text.addFilters}
            onPress={onFilterButtonPress}
            style={styles.filterButton}
            badgeValue={
              numTagFiltersSelected > 0 ? numTagFiltersSelected : null
            }
            testID="open-area-and-price-filters-button"
          />
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: filterBgColor
  },
  content: {
    marginTop: 16,
    paddingBottom: 12
  },
  contentFilters: {
    flexDirection: "row",
    justifyContent: "space-between",
    backgroundColor: filterButtonsBgColor,
    height: 48
  },
  filterButton: {
    flex: 1
  },
  dividerLine: {
    borderLeftWidth: StyleSheet.hairlineWidth,
    borderColor: whiteColor,
    opacity: 0.4
  }
});

export default FilterHeader;
