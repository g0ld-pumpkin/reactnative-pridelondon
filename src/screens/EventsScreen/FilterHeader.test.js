// @flow
import React from "react";
import { shallow } from "enzyme";
import FilterHeader from "./FilterHeader";
import type { Props as ComponentProps } from "./FilterHeader";
import FilterHeaderButton from "./FilterHeaderButton";
import FilterHeaderCategories from "./FilterHeaderCategories";
import ResetAllFiltersButton from "./ResetAllFiltersButton";

const render = (
  props: ComponentProps = {
    dateFilter: null,
    selectedCategories: new Set(),
    onFilterCategoriesPress: () => {},
    onFilterButtonPress: () => {},
    onDateFilterButtonPress: () => {},
    resetAllFiltersPress: () => {},
    numTagFiltersSelected: 0,
    scrollEventListToTop: () => {}
  }
) => shallow(<FilterHeader {...props} />);

describe("renders correctly", () => {
  it("with any date and any time", () => {
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: () => {}
    });
    expect(output).toMatchSnapshot();
  });

  it("with any date and any time (empty time set)", () => {
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: () => {}
    });
    expect(output).toMatchSnapshot();
  });

  it("with single date and single time", () => {
    const output = render({
      dateFilter: { startDate: "2018-02-02", endDate: "2018-02-02" },
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: () => {}
    });
    expect(output).toMatchSnapshot();
  });

  it("with date range and two times", () => {
    const output = render({
      dateFilter: {
        startDate: "2018-02-02",
        endDate: "2018-02-03"
      },
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: () => {}
    });
    expect(output).toMatchSnapshot();
  });

  it("with tag filters selected", () => {
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 2,
      scrollEventListToTop: () => {}
    });
    expect(output).toMatchSnapshot();
  });
});

describe("filter buttons", () => {
  it("calls onFilterCategoriesPress when users presses categories filter button", () => {
    const mock = jest.fn();
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: mock,
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: () => {}
    });
    output.find(FilterHeaderCategories).prop("onFilterPress")();

    expect(mock).toBeCalledWith();
  });

  it("calls onDateFilterButtonPress when users presses date filter button", () => {
    const mock = jest.fn();
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: mock,
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: () => {}
    });
    const button = output.find(FilterHeaderButton).at(0);
    button.simulate("press");

    expect(mock).toBeCalledWith();
  });

  it("calls onFilterButtonPress when users presses attribute filter button", () => {
    const mock = jest.fn();
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: mock,
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: () => {}
    });
    const button = output.find(FilterHeaderButton).at(1);
    button.simulate("press");

    expect(mock).toBeCalledWith();
  });

  it("calls scrollEventListToTop when users presses 'Reset all filters' button", () => {
    const mock = jest.fn();
    const output = render({
      dateFilter: null,
      selectedCategories: new Set(["Music"]),
      onFilterCategoriesPress: () => {},
      onFilterButtonPress: () => {},
      onDateFilterButtonPress: () => {},
      resetAllFiltersPress: () => {},
      numTagFiltersSelected: 0,
      scrollEventListToTop: mock
    });
    output
      .find(ResetAllFiltersButton)
      .props()
      .onPress();

    expect(mock).toBeCalledWith();
  });
});
