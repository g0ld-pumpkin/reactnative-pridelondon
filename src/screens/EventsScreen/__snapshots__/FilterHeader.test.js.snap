// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with any date and any time (empty time set) 1`] = `
<Component
  accessibilityTraits={
    Array [
      "header",
    ]
  }
  style={
    Object {
      "backgroundColor": "rgb(45, 47, 127)",
    }
  }
>
  <ContentPadding
    padding={Object {}}
    style={Object {}}
  >
    <Component>
      <ResetAllFiltersButton
        animationDelay={0}
        animationTime={300}
        onPress={[Function]}
        visible={false}
      />
      <Component
        style={
          Object {
            "marginTop": 16,
            "paddingBottom": 12,
          }
        }
        testID="event-filter-header"
      >
        <FilterHeaderCategories
          onFilterPress={[Function]}
          selectedCategories={Set {}}
        />
      </Component>
    </Component>
  </ContentPadding>
  <Component
    style={
      Object {
        "backgroundColor": "#20215a",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "space-between",
      }
    }
  >
    <FilterHeaderButton
      active={false}
      badgeValue={null}
      label="filter by date: Select dates"
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-date-filters-button"
      text="Select dates"
    />
    <Component
      style={
        Object {
          "borderColor": "#ffffff",
          "borderLeftWidth": 1,
          "opacity": 0.4,
        }
      }
    />
    <FilterHeaderButton
      active={false}
      badgeValue={null}
      label="Filter by area, price..."
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-area-and-price-filters-button"
      text="Filter by area, price..."
    />
  </Component>
</Component>
`;

exports[`renders correctly with any date and any time 1`] = `
<Component
  accessibilityTraits={
    Array [
      "header",
    ]
  }
  style={
    Object {
      "backgroundColor": "rgb(45, 47, 127)",
    }
  }
>
  <ContentPadding
    padding={Object {}}
    style={Object {}}
  >
    <Component>
      <ResetAllFiltersButton
        animationDelay={0}
        animationTime={300}
        onPress={[Function]}
        visible={false}
      />
      <Component
        style={
          Object {
            "marginTop": 16,
            "paddingBottom": 12,
          }
        }
        testID="event-filter-header"
      >
        <FilterHeaderCategories
          onFilterPress={[Function]}
          selectedCategories={Set {}}
        />
      </Component>
    </Component>
  </ContentPadding>
  <Component
    style={
      Object {
        "backgroundColor": "#20215a",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "space-between",
      }
    }
  >
    <FilterHeaderButton
      active={false}
      badgeValue={null}
      label="filter by date: Select dates"
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-date-filters-button"
      text="Select dates"
    />
    <Component
      style={
        Object {
          "borderColor": "#ffffff",
          "borderLeftWidth": 1,
          "opacity": 0.4,
        }
      }
    />
    <FilterHeaderButton
      active={false}
      badgeValue={null}
      label="Filter by area, price..."
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-area-and-price-filters-button"
      text="Filter by area, price..."
    />
  </Component>
</Component>
`;

exports[`renders correctly with date range and two times 1`] = `
<Component
  accessibilityTraits={
    Array [
      "header",
    ]
  }
  style={
    Object {
      "backgroundColor": "rgb(45, 47, 127)",
    }
  }
>
  <ContentPadding
    padding={Object {}}
    style={Object {}}
  >
    <Component>
      <ResetAllFiltersButton
        animationDelay={0}
        animationTime={300}
        onPress={[Function]}
        visible={true}
      />
      <Component
        style={
          Object {
            "marginTop": 16,
            "paddingBottom": 12,
          }
        }
        testID="event-filter-header"
      >
        <FilterHeaderCategories
          onFilterPress={[Function]}
          selectedCategories={Set {}}
        />
      </Component>
    </Component>
  </ContentPadding>
  <Component
    style={
      Object {
        "backgroundColor": "#20215a",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "space-between",
      }
    }
  >
    <FilterHeaderButton
      active={true}
      badgeValue={null}
      label="filter by date: 2 Feb - 3 Feb"
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-date-filters-button"
      text="2 Feb - 3 Feb"
    />
    <Component
      style={
        Object {
          "borderColor": "#ffffff",
          "borderLeftWidth": 1,
          "opacity": 0.4,
        }
      }
    />
    <FilterHeaderButton
      active={false}
      badgeValue={null}
      label="Filter by area, price..."
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-area-and-price-filters-button"
      text="Filter by area, price..."
    />
  </Component>
</Component>
`;

exports[`renders correctly with single date and single time 1`] = `
<Component
  accessibilityTraits={
    Array [
      "header",
    ]
  }
  style={
    Object {
      "backgroundColor": "rgb(45, 47, 127)",
    }
  }
>
  <ContentPadding
    padding={Object {}}
    style={Object {}}
  >
    <Component>
      <ResetAllFiltersButton
        animationDelay={0}
        animationTime={300}
        onPress={[Function]}
        visible={true}
      />
      <Component
        style={
          Object {
            "marginTop": 16,
            "paddingBottom": 12,
          }
        }
        testID="event-filter-header"
      >
        <FilterHeaderCategories
          onFilterPress={[Function]}
          selectedCategories={Set {}}
        />
      </Component>
    </Component>
  </ContentPadding>
  <Component
    style={
      Object {
        "backgroundColor": "#20215a",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "space-between",
      }
    }
  >
    <FilterHeaderButton
      active={true}
      badgeValue={null}
      label="filter by date: 2 Feb"
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-date-filters-button"
      text="2 Feb"
    />
    <Component
      style={
        Object {
          "borderColor": "#ffffff",
          "borderLeftWidth": 1,
          "opacity": 0.4,
        }
      }
    />
    <FilterHeaderButton
      active={false}
      badgeValue={null}
      label="Filter by area, price..."
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-area-and-price-filters-button"
      text="Filter by area, price..."
    />
  </Component>
</Component>
`;

exports[`renders correctly with tag filters selected 1`] = `
<Component
  accessibilityTraits={
    Array [
      "header",
    ]
  }
  style={
    Object {
      "backgroundColor": "rgb(45, 47, 127)",
    }
  }
>
  <ContentPadding
    padding={Object {}}
    style={Object {}}
  >
    <Component>
      <ResetAllFiltersButton
        animationDelay={0}
        animationTime={300}
        onPress={[Function]}
        visible={true}
      />
      <Component
        style={
          Object {
            "marginTop": 16,
            "paddingBottom": 12,
          }
        }
        testID="event-filter-header"
      >
        <FilterHeaderCategories
          onFilterPress={[Function]}
          selectedCategories={Set {}}
        />
      </Component>
    </Component>
  </ContentPadding>
  <Component
    style={
      Object {
        "backgroundColor": "#20215a",
        "flexDirection": "row",
        "height": 48,
        "justifyContent": "space-between",
      }
    }
  >
    <FilterHeaderButton
      active={false}
      badgeValue={null}
      label="filter by date: Select dates"
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-date-filters-button"
      text="Select dates"
    />
    <Component
      style={
        Object {
          "borderColor": "#ffffff",
          "borderLeftWidth": 1,
          "opacity": 0.4,
        }
      }
    />
    <FilterHeaderButton
      active={true}
      badgeValue={2}
      label="Filters"
      onPress={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      testID="open-area-and-price-filters-button"
      text="Filters"
    />
  </Component>
</Component>
`;
